### ISSUE TITLE
Gradient compression

### ISSUE THREAD

**Issue by vineeths96** [tensorflow/tensorflow#40468]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.2
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Feature: Gradient compression
Description: In distributed deep learning, communication of gradients often become a bottleneck. Gradient compression allows the gradients to be compressed so that the cost of communication is reduced. [Horovod](https://github.com/horovod/horovod) has a compression scheme built-in [here](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/compression.py). The compression can speed up the distributed training (proved by various research). Building such a feature to Tensorflow would facilitate research as well as development.

**Will this change the current api? How?**

**Who will benefit with this feature?**
People who are interested in large scale distributed training. 

**Any Other info.**

**Comment by vineeths96** [comment]
Pytorch is working on something similar [here](https://github.com/pytorch/pytorch/issues/39272).

Normally in a single worker system we can obtain the gradients using `compute_gradients()` function,  and make modifications to it, and then apply the gradients using `apply_gradients()` function. 

How does this change in the distributed setting? More specifically how can we access the gradients before and after the aggregation (or all_reduce)?

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
